src/projection.jl

```diff
@@ -141,16 +141,42 @@ end
 # dx::AbstractArray (when both are possible), or the reverse. So for now we just pass them through:
 (::ProjectTo{T})(dx::Tangent{<:T}) where {T} = dx
 
+#####
+##### A related utility which wants to live nearby
+#####
+
+"""
+    is_non_differentiable(x) == is_non_differentiable(typeof(x))
+
+Returns `true` if `x` is known from its type not to have derivatives, else `false`.
+
+Should mostly agree with whether `ProjectTo(x)` maps to `AbstractZero`,
+which is what the fallback method checks. The exception is that it will not look
+inside abstractly typed containers like `x = Any[true, false]`.
+"""
+is_non_differentiable(x) = is_non_differentiable(typeof(x))
+
+is_non_differentiable(::Type{<:Number}) = false
+is_non_differentiable(::Type{<:NTuple{N,T}}) where {N,T} = is_non_differentiable(T)
+is_non_differentiable(::Type{<:AbstractArray{T}}) where {T} = is_non_differentiable(T)
+
+function is_non_differentiable(::Type{T}) where {T}  # fallback
+    PT = Base._return_type(ProjectTo, Tuple{T})  # might be Union{} if unstable
```

Comment:
I kinda want to run this the other way around.
Can we make it so that the `is_non_differentiable` is the canonical one and `ProjectTo` falls back to that?

Reply:
That would make some sense. It's a slightly bigger API change I guess, as this is then what you ought to define (although I imagine 3rd-party non-diff types are pretty rare).

What this way around does do is re-use the logic for e.g. deciding whether all elements of a tuple are non-diff, hence the whole thing is. `ProjectTo(::Tuple)` already has to have code for recursing into things. If we reverse it, then `is_non_differentiable` will also want to have such code.